scipy/stats/_mannwhitneyu.py

```diff
@@ -0,0 +1,424 @@
+    '''Standardized MWU statistic'''
+    # Follows mannwhitneyu [2]
+    mu = n1 * n2 / 2
+    n = n1 + n2
+
+    # Tie correction according to [2]
+    tie_term = np.apply_along_axis(_tie_term, -1, ranks)
+    s = np.sqrt(n1*n2/12 * ((n + 1) - tie_term/(n*(n-1))))
+
+    # equivalent to using scipy.stats.tiecorrect
+    # T = np.apply_along_axis(stats.tiecorrect, -1, ranks)
+    # s = np.sqrt(T * n1 * n2 * (n1+n2+1) / 12.0)
+
+    numerator = U - mu
+
+    # Continuity correction.
+    # Because SF is always used to calculate the p-value, we can always
+    # _subtract_ 0.5 for the continuity correction. This always increases the
+    # p-value to accounts for the rest of the probability mass _at_ q = U.
+    if continuity:
+        numerator -= 0.5
+
+    # no problem evaluating the norm SF at an infinity
+    with np.errstate(divide='ignore', invalid='ignore'):
+        z = numerator / s
+    return z
+
+
+def _mwu_input_validation(x, y, use_continuity, alternative, axis, method):
+    ''' Input validation and standardization for mannwhitneyu '''
+    # Would use np.asarray_chkfinite, but infs are OK
+    x, y = np.atleast_1d(x), np.atleast_1d(y)
+    if np.isnan(x).any() or np.isnan(y).any():
+        raise ValueError('`x` and `y` must not contain NaNs.')
+    if np.size(x) == 0 or np.size(y) == 0:
+        raise ValueError('`x` and `y` must be of nonzero size.')
+
+    bools = {True, False}
+    if use_continuity not in bools:
+        raise ValueError(f'`use_continuity` must be one of {bools}.')
+
+    alternatives = {"two-sided", "less", "greater"}
+    alternative = alternative.lower()
+    if alternative not in alternatives:
+        raise ValueError(f'`alternative` must be one of {alternatives}.')
+
+    axis_int = int(axis)
+    if axis != axis_int:
+        raise ValueError('`axis` must be an integer.')
+
+    methods = {"asymptotic", "exact", "auto"}
+    method = method.lower()
+    if method not in methods:
+        raise ValueError(f'`method` must be one of {methods}.')
+
+    return x, y, use_continuity, alternative, axis_int, method
+
+
+def _tie_check(xy):
+    """Find any ties in data"""
+    _, t = np.unique(xy, return_counts=True, axis=-1)
+    return np.any(t != 1)
+
+
+def _mwu_choose_method(n1, n2, xy, method):
+    """Choose method 'asymptotic' or 'exact' depending on input size, ties"""
+
+    # if both inputs are large, asymptotic is OK
+    if n1 > 8 and n2 > 8:
+        return "asymptotic"
+
+    # if there are any ties, asymptotic is preferred
+    if np.apply_along_axis(_tie_check, -1, xy).any():
+        return "asymptotic"
+
+    return "exact"
+
+
+MannwhitneyuResult = namedtuple('MannwhitneyuResult', ('statistic', 'pvalue'))
+
+
+def mannwhitneyu(x, y, use_continuity=True, alternative="two-sided",
+                 axis=0, method="auto"):
+    r'''Perform the Mann-Whitney U rank test on two independent samples.
+
+    The Mann-Whitney U test is a nonparametric test of the null hypothesis
```

Comment:
Thanks; I've read your comments and taken a look at the email. I agree that there are simpler ways of stating the null hypothesis. When I wrote this, I had considered one of the other forms of the null hypothesis: that there is no difference in the _locations_ (mean/median) of the distributions, which is somewhat easier to write and think about. But that requires stronger assumptions about the difference between the two distributions, so I opted for the more general statement.

Considering your suggestion [below](https://github.com/scipy/scipy/pull/4933/files#r610907387), we could consider stating the null hypothesis that "the distributions underlying the samples are the same". That is definitely a simpler way of writing the null hypothesis, but then we also have to state the alternatives. Conover writes this all with symbols, e.g. H_0: P(X < Y) = 1/2, but I thought it would be good to stick with English, i.e. "stochastically less/greater", and I wanted to keep the language consistent between the descriptions of the null and alternatives... so this is what I ended up with.

Do you have a suggestion for simplifying the alternatives, too?

Reply:
Wikipedia writes "the probability of _X_ being greater than _Y_ is equal to the probability of _Y_ being greater than _X_."

By defining symbols _X_ and _Y_, we could get around having to repeat "the distribution underlying _x_" and "the distribution underlying _y_".

Then comparing the two:

"the probability of _X_ being greater than _Y_ is equal to the probability of _Y_ being greater than _X_."
"_X_ is neither stochastically less nor stochastically greater than _Y_"

I think I'd still prefer the latter because the terms "stochastically less" and "stochastically greater" are very useful for defining the alternatives.

Reply:
After reading a bit more about the test, I propose to state the null hypothesis using distribution functions, i.e.

H0: F = G

For the alternatives, we can write:

two-sided:  F(u) \ne G(u) for some u
less: The distribution underlying `x` is stochastically less than the distribution underlying `y`, i.e. F(u) > G(u) for all u.
greater: The distribution underlying `x` is stochastically greater than the distribution underlying `y`, i.e. F(u) < G(u) for all u.

In the Notes section, we could mention that there are other ways to write the hypothesis and refer to the paper https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2857732/pdf/nihms-185373.pdf

Rationale: 

- In the original paper (ref [1] in the docstring), the null hypothesis is also described as F = G
- The assumption F=G is crucial to derive the distribution of the test statistic
- We also keep the description of the alternative aligned to [1] while also mentioning distribution functions which is consistent with other GoF tests in SciPy

I hope it makes sense :)

Reply:
I tried to implement this in the latest commit. (Actually included it in the merge commit by accident, but this will need an interactive rebase anyway.) I agree that it's an improvement.

One suggestion for a possible final tweak is that we leave the other changes but not mention the cumulative distribution functions.

* People are confused by the ordering. X stochastically greater than Y (intuitive) is expressed F(u) < G(u)
* I don't think F(u) < G(u) is technically correct according to the reference you gave. It should be "F(u) <= G(u) with strict inequality for at least some u", right? Unfortunately, this is no more concise than "X is stochastically greater than Y".
* It requires the definition and use of mathematical symbols. We are discouraged from formatting them using LaTeX, so math ends up looking equally unprofessional in both the console and rendered documentation.

What do you think?

Reply:
Hi, I read the current version, it looks good. For the alternative 'less', there is a strict inequality in [1], in [5] it is <= for all x and < at least for one x. Not sure which one is correct :) I think the latter...
But I am also fine to just say 'stochastically less'

Reply:
OK, implemented [here](https://github.com/scipy/scipy/pull/4933/files#r615412274). Please commit that (and [this](https://github.com/scipy/scipy/pull/4933/files#r615412490) typo fix above) if it looks good.